File: frontend/src/i18n/uzTranslit.ts
/**
 * Узбекская латиница → узбекская кириллица (локаль `uz_cyrl`).
 *
 * Порт `uk_management_bot/utils/uz_translit.py` (Фаза 3): фронтовая
 * локаль uz_cyrl = транслитерация всего uz.json + поверх ручные правки
 * `locales/uz_cyrl.json` (только там, где правило ошибается). Так новые
 * ключи uz сразу появляются и в кириллице, без смеси письменностей.
 *
 * Правила (официальная орфография 1995 г. в обратную сторону):
 * - o‘ / g‘ (апостроф ‘ ʻ ' ’ `) → ў / ғ; sh → ш, ch → ч, ng → нг;
 * - yo / yu / ya → ё / ю / я; ye → е в начале слова и после гласной, иначе ъе;
 *   одиночная y → й;
 * - e → э в начале слова и после гласной, иначе е;
 * - q → қ, h → ҳ, x → х, j → ж; ts перед «iy»/«io» → ц, прочие ts — т+с;
 * - тутуқ белгиси (' ’ ʼ) после гласной перед буквой → ъ; s'h → сҳ;
 *   апостроф после согласной — кавычка, остаётся.
 *
 * Не трогаются: плейсхолдеры i18next `{{…}}` и `{…}`, вложенные `$t(…)`,
 * HTML-теги и сущности, URL, @username, #теги, /команды, `код`,
 * snake_case-идентификаторы, аббревиатуры и бренды из ABBREVIATIONS.
 */

// Латинские аббревиатуры/бренды, которые в кириллице остаются латиницей.
const ABBREVIATIONS = [
  'TWA', 'ID', 'SMS', 'QR', 'PDF', 'API', 'URL', 'AI', 'OK', 'GPS', 'TV',
  'JPG', 'PNG', 'DOCX', 'DOC', 'MB', 'KB', 'HH:MM',
  'Telegram', 'InfraSafe', 'PROFK', 'Excel', 'CSV', 'XLSX', 'Wi-Fi', 'Email', 'E-mail', 'email',
]

const PROTECTED_RE = new RegExp(
  [
    String.raw`\{\{[^{}]*\}\}`, // плейсхолдеры i18next
    String.raw`\{[^{}]*\}`, // плейсхолдеры {name}
    String.raw`\$t\([^)]*\)`, // вложенные ключи i18next
    String.raw`<[^<>]+>`, // HTML-теги вместе с атрибутами
    String.raw`&(?:[A-Za-z]+|#\d+|#x[0-9A-Fa-f]+);`, // HTML-сущности
    String.raw`(?:https?://|www\.)[^\s<>"']+`, // URL
    String.raw`(?<![\w/])t\.me/[^\s<>"']+`,
    String.raw`(?<!\w)@\w+`, // @username
    String.raw`(?<!\w)#\w+`, // #теги
    String.raw`(?<![\w/])/[A-Za-z_]+`, // /команды
    String.raw`(?<![A-Za-z])` + '`[^`\\n]+`', // `код` (не o`/g`)
    String.raw`\b[A-Za-z]+(?:_[A-Za-z0-9]+)+(?:=[A-Za-z0-9]+)?\b`, // snake_case
    String.raw`(?<![A-Za-z])(?:` + ABBREVIATIONS.map((a) => a.replace(/[-:]/g, (c) => `\\${c}`)).join('|') + String.raw`)(?![A-Za-z])`,
  ].join('|'),
  'g',
)

const SINGLE: Record<string, string> = {
  a: 'а', b: 'б', c: 'ц', d: 'д', f: 'ф', g: 'г', h: 'ҳ', i: 'и', j: 'ж', k: 'к',
  l: 'л', m: 'м', n: 'н', o: 'о', p: 'п', q: 'қ', r: 'р', s: 'с', t: 'т', u: 'у',
  v: 'в', w: 'в', x: 'х', y: 'й', z: 'з',
}
const Y_VOWELS: Record<string, string> = { o: 'ё', u: 'ю', a: 'я' }
const VOWELS = new Set('aeiouаеёиоуэюяў')
// Апостроф-модификатор в o‘/g‘; ‘ и ʻ однозначны, остальные — только перед буквой.
const MODIFIERS_STRICT = new Set(['‘', 'ʻ'])
const MODIFIERS_LOOSE = new Set(["'", '’', '`'])
// Тутуқ белгиси (разделительный знак).
const TUTUQ = new Set(["'", '’', 'ʼ'])

function isLetter(ch: string | undefined): boolean {
  return !!ch && /\p{L}/u.test(ch)
}

function isUpper(ch: string): boolean {
  return ch !== ch.toLowerCase() && ch === ch.toUpperCase()
}

function withCase(src: string, cyr: string): string {
  return isUpper(src) ? cyr.toUpperCase() : cyr
}

function isModifier(s: string, i: number): boolean {
  if (i >= s.length) return false
  if (MODIFIERS_STRICT.has(s[i])) return true
  return MODIFIERS_LOOSE.has(s[i]) && isLetter(s[i + 1])
}

function wordStartOrAfterVowel(s: string, i: number): boolean {
  if (i === 0) return true
  const prev = s[i - 1]
  return !isLetter(prev) || VOWELS.has(prev.toLowerCase())
}

/** Одна единица транслитерации: [кириллица, сколько символов съедено]. */
function step(s: string, i: number): [string, number] {
  const ch = s[i]
  const low = ch.toLowerCase()
  const nxt = (s[i + 1] ?? '').toLowerCase()

  if ((low === 'o' || low === 'g') && isModifier(s, i + 1)) return [withCase(ch, low === 'o' ? 'ў' : 'ғ'), 2]
  if (low === 's' && nxt === 'h') return [withCase(ch, 'ш'), 2]
  if (low === 'c' && nxt === 'h') return [withCase(ch, 'ч'), 2]
  if (low === 't' && nxt === 's' && ['iy', 'io'].includes(s.slice(i + 2, i + 4).toLowerCase())) return [withCase(ch, 'ц'), 2]
  if (low === 'y' && nxt in Y_VOWELS && !(nxt === 'o' && isModifier(s, i + 2))) return [withCase(ch, Y_VOWELS[nxt]), 2]
  if (low === 'y' && nxt === 'e') {
    return wordStartOrAfterVowel(s, i) ? [withCase(ch, 'е'), 2] : ['ъ' + withCase(ch, 'е'), 2]
  }
  if (low === 'e') return [withCase(ch, wordStartOrAfterVowel(s, i) ? 'э' : 'е'), 1]
  if (TUTUQ.has(ch) && i > 0 && isLetter(s[i - 1]) && nxt && isLetter(s[i + 1])) {
    const prev = s[i - 1].toLowerCase()
    if (prev === 's' && nxt === 'h') return ['', 1] // is'hoq → исҳоқ
    if (VOWELS.has(prev) || ch === 'ʼ') return ['ъ', 1]
    return [ch, 1] // после согласной — кавычка ('Yakunlash'ni)
  }
  if (low in SINGLE) return [withCase(ch, SINGLE[low]), 1]
  return [ch, 1]
}

function translitPlain(s: string): string {
  let out = ''
  let i = 0
  while (i < s.length) {
    const [piece, used] = step(s, i)
    out += piece
    i += used
  }
  return out
}

/** Транслитерировать строку, не трогая защищённые фрагменты. */
export function toCyrillic(text: string): string {
  let out = ''
  let pos = 0
  for (const m of text.matchAll(PROTECTED_RE)) {
    const start = m.index ?? 0
    out += translitPlain(text.slice(pos, start)) + m[0]
    pos = start + m[0].length
  }
  return out + translitPlain(text.slice(pos))
}

/** Текст без защищённых фрагментов (для гейта «латиница не осталась»). */
export function stripProtected(text: string): string {
  return text.replace(PROTECTED_RE, ' ')
}

/** Новая структура: все строковые значения в кириллице, ключи как есть. */
export function transliterateTree<T>(data: T): T {
  if (Array.isArray(data)) return data.map((v) => transliterateTree(v)) as T
  if (data && typeof data === 'object') {
    return Object.fromEntries(
      Object.entries(data as Record<string, unknown>).map(([k, v]) => [k, transliterateTree(v)]),
    ) as T
  }
  if (typeof data === 'string') return toCyrillic(data) as T
  return data
}

type Tree = Record<string, unknown>

/** Новый объект: `base` с наложенными поверх `overrides` (рекурсивно). */
export function deepMerge(base: Tree, overrides: Tree): Tree {
  const merged: Tree = { ...base }
  for (const [key, value] of Object.entries(overrides)) {
    const cur = merged[key]
    merged[key] =
      value && typeof value === 'object' && !Array.isArray(value) && cur && typeof cur === 'object' && !Array.isArray(cur)
        ? deepMerge(cur as Tree, value as Tree)
        : value
  }
  return merged
}

File: uk_management_bot/utils/uz_translit.py
"""Узбекская латиница → узбекская кириллица (локаль бота ``uz_cyrl``).

Локаль ``uz_cyrl`` не поддерживается вручную: ``utils/helpers.load_locale``
строит её из ``uz.json`` этой транслитерацией (новые ключи в uz.json сразу
появляются и в кириллице). Правки для слов, где правило ошибается, — в
``config/locales/uz_cyrl.overrides.json``.

Правила (официальная орфография 1995 г. в обратную сторону):
- o‘ / g‘ (апостроф ‘ ʻ ' ’ `) → ў / ғ; sh → ш, ch → ч, ng → нг;
- yo / yu / ya → ё / ю / я; ye → е в начале слова и после гласной, иначе ъе
  (obyekt → объект); одиночная y → й;
- e → э в начале слова и после гласной, иначе е;
- q → қ, h → ҳ, x → х, j → ж; ts перед «iy»/«io» → ц (operatsiya → операция,
  но muvaffaqiyatsiz → муваффақиятсиз); прочие ts — т+с;
- тутуқ белгиси (' ’ ʼ) после гласной перед буквой → ъ (ta’mir → таъмир);
  s'h → сҳ; апостроф после согласной — это кавычка, остаётся как есть.
- регистр: Sh / SH → Ш, прочее — по регистру первой буквы.

Не трогаются: плейсхолдеры ``{...}``, HTML-теги и сущности, URL, ``@username``,
``#теги``, ``/команды``, аббревиатуры из ``ABBREVIATIONS``, эмодзи и уже
кириллический текст.
"""
import re
from typing import Any, Dict

# Латинские аббревиатуры/бренды, которые в кириллице остаются латиницей.
ABBREVIATIONS = (
    "TWA", "ID", "SMS", "QR", "PDF", "API", "URL", "AI", "OK", "GPS", "TV",
    "JPG", "PNG", "DOCX", "DOC", "MB", "KB", "HH:MM",
)

_PROTECTED_RE = re.compile(
    "|".join(
        (
            r"\{[^{}]*\}",  # плейсхолдеры str.format
            r"<[^<>]+>",  # HTML-теги вместе с атрибутами
            r"&(?:[A-Za-z]+|#\d+|#x[0-9A-Fa-f]+);",  # HTML-сущности
            r"(?:https?://|www\.)[^\s<>\"']+",  # URL
            r"(?<![\w/])t\.me/[^\s<>\"']+",
            r"(?<!\w)@\w+",  # @username
            r"(?<!\w)#\w+",  # #теги
            r"(?<![\w/])/[A-Za-z_]+",  # /команды
            r"(?<![A-Za-z])`[^`\n]+`",  # `код` (не o`/g` — перед ` нет буквы)
            r"\b[A-Za-z]+(?:_[A-Za-z0-9]+)+(?:=[A-Za-z0-9]+)?\b",  # snake_case-идентификаторы
            r"\b(?:" + "|".join(ABBREVIATIONS) + r")\b",
        )
    )
)

_SINGLE = {
    "a": "а", "b": "б", "c": "ц", "d": "д", "f": "ф", "g": "г", "h": "ҳ",
    "i": "и", "j": "ж", "k": "к", "l": "л", "m": "м", "n": "н", "o": "о",
    "p": "п", "q": "қ", "r": "р", "s": "с", "t": "т", "u": "у", "v": "в",
    "w": "в", "x": "х", "y": "й", "z": "з",
}
_Y_VOWELS = {"o": "ё", "u": "ю", "a": "я"}
_VOWELS = set("aeiouаеёиоуэюяў")
# Апостроф-модификатор в o‘/g‘; ‘ и ʻ однозначны, остальные — только перед буквой.
_MODIFIERS_STRICT = {"‘", "ʻ"}
_MODIFIERS_LOOSE = {"'", "’", "`"}
# Тутуқ белгиси (разделительный знак).
_TUTUQ = {"'", "’", "ʼ"}


def _case(src: str, cyr: str) -> str:
    return cyr.upper() if src.isupper() else cyr


def _is_letter(ch: str) -> bool:
    return ch.isalpha()


def _is_modifier(s: str, i: int) -> bool:
    """s[i] — апостроф o‘/g‘?"""
    if i >= len(s):
        return False
    if s[i] in _MODIFIERS_STRICT:
        return True
    return s[i] in _MODIFIERS_LOOSE and i + 1 < len(s) and _is_letter(s[i + 1])


def _word_start_or_after_vowel(s: str, i: int) -> bool:
    if i == 0:
        return True
    prev = s[i - 1]
    return not _is_letter(prev) or prev.lower() in _VOWELS


def _step(s: str, i: int) -> tuple:
    """Одна единица транслитерации: (кириллица, сколько символов съедено)."""
    ch = s[i]
    low = ch.lower()
    nxt = s[i + 1].lower() if i + 1 < len(s) else ""

    if low in ("o", "g") and _is_modifier(s, i + 1):
        return _case(ch, "ў" if low == "o" else "ғ"), 2
    if low == "s" and nxt == "h":
        return _case(ch, "ш"), 2
    if low == "c" and nxt == "h":
        return _case(ch, "ч"), 2
    if low == "t" and nxt == "s" and s[i + 2:i + 4].lower() in ("iy", "io"):
        return _case(ch, "ц"), 2
    if low == "y" and nxt in _Y_VOWELS and not (nxt == "o" and _is_modifier(s, i + 2)):
        return _case(ch, _Y_VOWELS[nxt]), 2
    if low == "y" and nxt == "e":
        if _word_start_or_after_vowel(s, i):
            return _case(ch, "е"), 2
        return "ъ" + _case(ch, "е"), 2
    if low == "e":
        return _case(ch, "э" if _word_start_or_after_vowel(s, i) else "е"), 1
    if ch in _TUTUQ and i > 0 and _is_letter(s[i - 1]) and nxt and _is_letter(s[i + 1]):
        prev = s[i - 1].lower()
        if prev == "s" and nxt == "h":
            return "", 1  # is'hoq → исҳоқ: разделяет с и ҳ
        if prev in _VOWELS or ch == "ʼ":
            return "ъ", 1
        return ch, 1  # после согласной — кавычка ('Yakunlash'ni)
    if low in _SINGLE:
        return _case(ch, _SINGLE[low]), 1
    return ch, 1


def _translit_plain(s: str) -> str:
    out = []
    i = 0
    while i < len(s):
        piece, used = _step(s, i)
        out.append(piece)
        i += used
    return "".join(out)


def to_cyrillic(text: str) -> str:
    """Транслитерировать строку, не трогая защищённые фрагменты."""
    parts = []
    pos = 0
    for match in _PROTECTED_RE.finditer(text):
        parts.append(_translit_plain(text[pos:match.start()]))
        parts.append(match.group(0))
        pos = match.end()
    parts.append(_translit_plain(text[pos:]))
    return "".join(parts)


def strip_protected(text: str) -> str:
    """Текст без защищённых фрагментов (для гейта «латиница не осталась»)."""
    return _PROTECTED_RE.sub(" ", text)


def transliterate_tree(data: Any) -> Any:
    """Новая структура: все строковые значения в кириллице, ключи как есть."""
    if isinstance(data, dict):
        return {key: transliterate_tree(value) for key, value in data.items()}
    if isinstance(data, list):
        return [transliterate_tree(value) for value in data]
    if isinstance(data, str):
        return to_cyrillic(data)
    return data


def deep_merge(base: Dict[str, Any], overrides: Dict[str, Any]) -> Dict[str, Any]:
    """Новый словарь: ``base`` с наложенными поверх ``overrides`` (рекурсивно)."""
    merged = dict(base)
    for key, value in overrides.items():
        if isinstance(value, dict) and isinstance(merged.get(key), dict):
            merged[key] = deep_merge(merged[key], value)
        else:
            merged[key] = value
    return merged
